File: go/cote/GJM/도시가_모두_연결되는_시점.go
package main

import (
	"fmt"
	"math/rand"
	"sort"
)

const (
	TEST_CASE = 10
	N = 10
	MAX_DELAY = 100
)

type Edge struct {
	From int
	To int
	Delay int
}

type Edges []Edge

func (e Edges) Len() int {
	return len(e)
}

func (e Edges) Less(i, j int) bool {
	return e[i].Delay < e[j].Delay
}

func (e Edges) Swap(i, j int) {
	e[i], e[j] = e[j], e[i]
}

type DisjointSet interface {
	Init(n int)
	Find(u int) int
	Merge(u, v int)
}

type TreeSet struct {
	Parent []int 
	Rank []int
	Size int
}

func (s *TreeSet) Init(n int){
	s.Parent = make([]int, n)
	s.Rank = make([]int, n)
	s.Size = n

	for i, _ := range s.Parent {
		s.Parent[i] = i 
	}
}

func (s *TreeSet) Find(u int) int {
	if s.Parent[u] == u {
		return u
	}

	s.Parent[u] = s.Find(s.Parent[u])

	return s.Parent[u]
}

func (s *TreeSet) Merge(u, v int) {
	u = s.Find(u)
	v = s.Find(v)

	if u == v {
		return
	}

	if s.Rank[u] < s.Rank[v] {
		u, v = v, u
	}
	s.Parent[v] = u
	
	if s.Rank[u] == s.Rank[v] {
		s.Rank[u]++ 
	} 
	s.Size--
}

func initialize(random *rand.Rand) (n int, edges []Edge) {
	n = random.Intn(N - 3) + 3
	edges = make([]Edge, n)

	for i, _ := range edges {
		for j := 0; j < n; j++ {
			if i == j || random.Float32() < 0.75 {
				continue
			}
			edges = append(edges, Edge{i, j, random.Intn(MAX_DELAY)})
		}
	}

	sort.Sort(Edges(edges))
	return 
}

func solve(n int, edges []Edge) int {
	set := new(TreeSet)
	set.Init(n)

	for _, edge := range edges{
		i, j, delay := edge.From, edge.To, edge.Delay

		if set.Size == 1 {
			return delay
		}

		set.Merge(i, j)
	}

	return -1
}

func main() {
	random := rand.New(rand.NewSource(43))

	for it := 0; it < N; it++ {
		n, edges := initialize(random)
		fmt.Println(solve(n, edges))
	}
}

File: c++/cote/GJM/도시가_모두_연결되는_시점.cpp
#include <algorithm>
#include <initializer_list>
#include <iostream>
#include <random>
#define N 10
#define TEST_CASES 10
#define MAX_TIME 100

using namespace std;

struct DisjointSet {
  vector<int> parent, rank;
  int size;

  DisjointSet(int n) : parent(n), rank(n - 1) {
    for (int i = 0; i < n; i++) {
      parent[i] = i;
    }
    size = n;
  }

  int find(int u) {
    if (u == parent[u]) return u;
    return parent[u] = find(parent[u]);
  }

  void merge(int u, int v) {
    u = find(u);
    v = find(v);

    if (u == v) return;
    if (rank[u] > rank[v]) swap(u, v);

    parent[u] = v;
    if (rank[u] == rank[v]) rank[u]++;
    size--;
  }
};

vector<vector<int>>* initialize(default_random_engine& gen, int n) {
  vector<vector<int>>* ret = new vector<vector<int>>();
  uniform_int_distribution<int> dis2(MAX_TIME);

  for (int i = 0; i < n; i++) {
    for (int j = i; j < n; j++) {
      ret->push_back(vector<int>{i, j, dis2(gen)});
    }
  }

  sort(begin(*ret), end(*ret),
       [](vector<int> a, vector<int> b) { return a[2] < b[2]; });

  for_each(begin(*ret), end(*ret), [](vector<int>& a) { a.pop_back(); });

  return ret;
}

int solve(vector<vector<int>>& edges, int n) {
  DisjointSet set(n);

  for (int t = 0; t < edges.size(); t++) {
    if (set.size == 1) {
      return t;
    }

    int i = edges[t][0];
    int j = edges[t][1];

    set.merge(i, j);
  }

  return -1;
}

int main(void) {
  default_random_engine gen(43);
  uniform_int_distribution<int> dis1(1, N);

  for (int it = 0; it < TEST_CASES; it++) {
    int n = dis1(gen);
    vector<vector<int>> edges = *(initialize(gen, n));
    cout << solve(edges, n) << endl;
  }

  return 0;
}